src/espidf.rs

```diff
@@ -102,6 +102,206 @@ impl Tools {
+    name: String,
+    /// url to obtain the Tool as an compressed binary
+    url: String,
+    /// version of the tool in no particular format
+    versions: String,
+    /// hash of the compressed file
+    sha256: String,
+    /// size of the compressed file
+    size: u64,
+    /// Base absolute install dir as absolute Path
+    install_dir: PathBuf,
+    /// Path relative to install dir
+    export_path: PathBuf,
+    /// Command path and args that printout the current version of the tool
+    /// - First element is the relative path to the command
+    /// - Every other element represents a arg given to the cmd
+    version_cmd_args: Vec<String>,
+    /// regex to extract the version returned by the version_cmd
+    version_regex: String,
+}
+
+impl Tool {
+    /// Test if the tool is installed correctly
+    fn test(&self) -> bool {
+        let tool_path = self.abs_export_path();
+
+        // if path does not exist -> tool is not installed
+        if !tool_path.exists() {
+            return false;
+        }
+        log::debug!(
+            "Run cmd: {:?} to get current tool version",
+            self.test_command(),
+        );
+
+        let output = self
+            .test_command()
+            .output()
+            .unwrap_or_else(|_| panic!("Failed to run command: {:?}", self.test_command()));
+
+        let regex = regex::Regex::new(&self.version_regex).expect("Invalid regex pattern provided");
+
+        if let Some(capture) = regex.captures(&String::from_utf8_lossy(&output.stdout)) {
+            if let Some(var) = capture.get(0) {
+                log::debug!("Match: {:?}, Version: {:?}", &var.as_str(), &self.versions);
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// get the absolute PATH
+    fn abs_export_path(&self) -> PathBuf {
+        self.install_dir.join(self.export_path.as_path())
+    }
+
+    /// Creates a Command that will echo back the current version of the tool
+    ///
+    /// Since Command is non clonable this helper is provided
+    fn test_command(&self) -> Command {
+        let cmd_abs_path = self
+            .abs_export_path()
+            .join(self.version_cmd_args[0].clone());
+
+        let mut version_cmd = std::process::Command::new(cmd_abs_path);
+        version_cmd.args(self.version_cmd_args[1..].iter().cloned());
+        version_cmd
+    }
+}
+
+/// Parsing a provided tools.json file, and return a Vec<Tool> representing a Tool version of the wanted tools
+fn parse_into_tools(
+    tools_wanted: Vec<&str>,
+    tools_json_file: PathBuf,
+    install_dir: PathBuf,
+) -> anyhow::Result<Vec<Tool>> {
+    let mut tools: Vec<Tool> = Vec::new();
+
+    let os_key = get_os_target_key().unwrap();
+
+    let mut tools_string = String::new();
+    let mut tools_file = std::fs::File::open(tools_json_file)?;
+
+    tools_file.read_to_string(&mut tools_string)?;
+
+    let parsed_file = serde_json::from_str::<serde_json::Value>(&tools_string)?;
```

Comment:
Why are you parsing the `tools.json` file as an untyped map?
I would assume it would still make sense to have an internal struct named `ToolsJson` or suchlike, which is strongly typed?

Reply:
Yeah, i was a bit lazy here. I think there is not to much benefit here for creating a separate strongly typed version over the raw serialized data ( skipping stuff that are not needed), but i can implement that so no problem there.

Reply:
Up to you, but it would lower the amounts of `unwrap`s in this code considerably, I think.
If you decide not to, just merge.

Reply:
ok i addressed that know, shooting with big guns now. :gun: In the process i found out multiple problems in the json schema. No unwraps anymore ( at least as far as the schema is concerned)